simplified-books-borrowing/src/main/java/org/nypl/simplified/books/borrowing/internal/BorrowLoanCreate.kt

```diff
@@ -0,0 +1,309 @@
+import org.nypl.simplified.opds.core.OPDSAvailabilityMatcherType
+import org.nypl.simplified.opds.core.OPDSAvailabilityOpenAccess
+import org.nypl.simplified.opds.core.OPDSAvailabilityRevoked
+import org.nypl.simplified.opds.core.OPDSParseException
+import java.io.ByteArrayInputStream
+import java.io.InputStream
+import java.net.URI
+
+/**
+ * A task that creates an OPDS loan by hitting an acquisition URI.
+ */
+
+class BorrowLoanCreate private constructor() : BorrowSubtaskType {
+
+  companion object : BorrowSubtaskFactoryType {
+    override val name: String
+      get() = "Create OPDS Loan"
+
+    override fun createSubtask(): BorrowSubtaskType {
+      return BorrowLoanCreate()
+    }
+
+    override fun isApplicableFor(
+      type: MIMEType,
+      target: URI?
+    ): Boolean {
+      for (opdsType in allOPDSFeeds) {
+        if (MIMECompatibility.isCompatibleStrictWithoutAttributes(opdsType, type)) {
+          return true
+        }
+      }
+      return false
+    }
+  }
+
+  override fun execute(context: BorrowContextType) {
+    context.taskRecorder.beginNewStep("Creating an OPDS loan...")
+
+    try {
+      context.bookLoanIsRequesting("Requesting a loan...")
+
+      val currentURI = context.currentURICheck()
+      context.taskRecorder.beginNewStep("Using $currentURI to create a loan...")
+      context.taskRecorder.addAttribute("Loan URI", currentURI.toString())
+      context.checkCancelled()
+
+      val request =
+        context.httpClient.newRequest(currentURI)
+          .setMethod(Put(ByteArray(0), applicationOctetStream))
+          .setAuthorization(authorizationOf(context.account))
+          .build()
+
+      return request.execute().use { response ->
+        when (val status = response.status) {
+          is LSHTTPResponseStatus.Responded.OK ->
+            this.handleOKRequest(context, currentURI, status)
+          is LSHTTPResponseStatus.Responded.Error ->
+            this.handleHTTPError(context, status)
+          is LSHTTPResponseStatus.Failed ->
+            this.handleHTTPFailure(context, status)
+        }
+      }
+    } catch (e: BorrowSubtaskFailed) {
+      context.bookLoanFailed()
+      throw e
+    }
+  }
+
+  private fun handleHTTPFailure(
+    context: BorrowContextType,
+    status: LSHTTPResponseStatus.Failed
+  ) {
+    context.taskRecorder.currentStepFailed(
+      message = status.exception.message ?: "Exception raised during connection attempt.",
+      errorCode = httpConnectionFailed,
+      exception = status.exception
+    )
+    throw BorrowSubtaskFailed()
+  }
+
+  private fun handleHTTPError(
+    context: BorrowContextType,
+    status: LSHTTPResponseStatus.Responded.Error
+  ) {
+    context.taskRecorder.addAttributes(BorrowHTTP.problemReportAsAttributes(status.problemReport))
+
+    val report = status.problemReport
+    if (report != null) {
+      if (report.type == "http://librarysimplified.org/terms/problem/loan-already-exists") {
```

Comment:
Possible for this to ever be an `https` type and break this equality check?

Reply:
Don't think so. It's similar to a namespace identifier in XML in that it's a URL but it doesn't actually describe a real resource. I wish we'd used URNs for these!

Reply:
Perfect. Just checking!